### PR TITLE
support plugin http execution on remote agents

### PR DESCRIPTION
**Pull request by mchugh19**

### CODE REVIEW

File: src/main/java/com/waytta/Builds.java [hunk 1]
```diff
@@ -70,13 +74,13 @@ public static void addArgumentsToSaltFunction(String myarguments, JSONObject sal
     }
     
 
-    public static JSONArray runBlockingBuild(Run build, JSONArray returnArray, String myservername, 
-    		String token, JSONObject saltFunc, TaskListener listener, int pollTime, int minionTimeout) throws InterruptedException {
+    public static JSONArray runBlockingBuild(Launcher launcher, Run build, JSONArray returnArray, String myservername,
```
Comment (oleg-nenashev) [inline src/main/java/com/waytta/Builds.java]: It formally breaks the binary compatibility. It is up to the plugin maintainer to decide if it is an issue or not in the case of this plugin

File: src/main/java/com/waytta/SaltAPIBuilder.java [hunk 2]
```diff
@@ -356,7 +359,7 @@ public static FormValidation doTestConnection(
                 @QueryParameter String servername,
                 @QueryParameter String credentialsId,
                 @QueryParameter String authtype,
-        		@AncestorInPath Item project) {            	
+        		@AncestorInPath Item project) throws InterruptedException, IOException {
```
Comment (oleg-nenashev) [inline src/main/java/com/waytta/SaltAPIBuilder.java]: It is a form validation method. In the case of these exceptions you will get a pretty bad UI. You would rather want to catch exceptions and return them in `FormValidation.error`

File: src/main/java/com/waytta/SaltAPIBuilder.java [hunk 3]
```diff
@@ -372,10 +375,12 @@ public static FormValidation doTestConnection(
             if (usedCredential == null) {
                 return FormValidation.error("CredentialId error: no credential found with given ID.");
             }
+            
+            Launcher launcher = Hudson.getInstance().createLauncher(TaskListener.NULL);
```
Comment (oleg-nenashev) [inline src/main/java/com/waytta/SaltAPIBuilder.java]: `TaskListener.NULL` suppresses all log entries. It should go to the System log at least

File: src/main/java/com/waytta/SaltAPIStep.java [hunk 4]
```diff
@@ -164,7 +166,7 @@ public FormValidation doTestConnection(
                 @QueryParameter String servername,
                 @QueryParameter String credentialsId,
                 @QueryParameter String authtype,
-                @AncestorInPath Item project) {
+                @AncestorInPath Item project) throws InterruptedException, IOException {
```
Comment (oleg-nenashev) [inline src/main/java/com/waytta/SaltAPIStep.java]: same as above

File: src/main/java/com/waytta/Utils.java [hunk 5]
```diff
@@ -34,76 +36,9 @@
 public class Utils {
     private static final String RETCODE_FIELD_NAME = "retcode";
 
-    // Thinger to connect to saltmaster over rest interface
-    public static JSONObject getJSON(String targetURL, JSONObject urlParams, String auth) {
```
Comment (oleg-nenashev) [inline src/main/java/com/waytta/Utils.java]: breaks compatibility as well. Not sure if it is actually required

File: src/main/java/com/waytta/saltAPI.java [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+    }
+    
+    public JSONObject call() throws IOException {
+        HttpURLConnection connection = null;
+        JSONObject responseJSON = new JSONObject();
+
+        try {
+            // Create connection
+            URL url = new URL(targetURL);
+            connection = (HttpURLConnection) url.openConnection();
```
Comment (oleg-nenashev) [inline src/main/java/com/waytta/saltAPI.java]: Possible Class Cast exception if a user specifies another protocol (e.g. "file:///" or "jar://" even if it id weird)

File: src/main/java/com/waytta/SaltAPIBuilder.java (hunk 3 above)
Comment (oleg-nenashev) [inline src/main/java/com/waytta/SaltAPIBuilder.java]: Also `Jenkins.getInstance()`

File: src/main/java/com/waytta/SaltAPIBuilder.java (hunk 2 above)
Comment (mchugh19) [inline src/main/java/com/waytta/SaltAPIBuilder.java]: Yep. Good suggestion

File: src/main/java/com/waytta/Builds.java (hunk 1 above)
Comment (mchugh19) [inline src/main/java/com/waytta/Builds.java]: It does, but I don't see any other users of these functions so it's worth the update

File: src/main/java/com/waytta/SaltAPIBuilder.java (hunk 3 above)
Comment (mchugh19) [inline src/main/java/com/waytta/SaltAPIBuilder.java]: I'm not sure how to instantiate system log from here. Any ideas?

File: src/main/java/com/waytta/saltAPI.java (hunk 6 above)
Comment (mchugh19) [inline src/main/java/com/waytta/saltAPI.java]: Looks like the catch should still catch that too.

**Pull request merged**